google/cloud/documentai_toolbox/wrappers/document_revision.py

```diff
@@ -0,0 +1,346 @@
+        base_shards.append(doc)
+
+    for byte in revision_bytes:
+        revision_shards.append(pb.FromString(byte))
+
+    return base_shards, revision_shards
+
+
+def _modify_docproto(entities):
+    entities_array = []
+    history = []
+    for e in entities:
+        if (
+            e.provenance.type_ == _OP_TYPE.ADD
+            or e.provenance.type_ == _OP_TYPE.OPERATION_TYPE_UNSPECIFIED
+        ):
+            entities_array.append(Entity(documentai_entity=e))
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                }
+            )
+        elif e.provenance.type_ == _OP_TYPE.REMOVE:
+            entity = entities_array.pop(int(e.id) - 1)
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": entity.type_,
+                    "original_text": entity.mention_text,
+                }
+            )
+            del entity
+        elif e.provenance.type_ == _OP_TYPE.REPLACE:
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                    "replace_type": e.type_,
+                    "replace_text": e.mention_text,
+                }
+            )
+            entities_array[int(e.id) - 1].replace(documentai_entity=e)
+    return entities_array, history
+
+
+def _get_revised_documents(revision: documentai.Document):
+    """
+    Given the immutable_document and revisions this function
+    will create Document objects correlating to the provenance in revision
+
+        For Example :
+            immutable_document = {text:"blah", pages:{documentai.Document.Page}}
+            revisions = [
+                Revision(rev_id=a1,doc_shard,rev_index,rev,parents=[]),
+                Revision(rev_id=a2,doc_shard,rev_index,rev,parents=[a1])
+            ]
+
+            This will return :
+                [8aba4bb1ec12db8a:{Document1},8aba4bb1ec12d275:{Document2}]
+                    Where
+                        Document1 = {pages:List[Page],entities:List[Entity],revision_nodes:List[RevisionNodes]...}
+                        Document2 = {pages:List[Page],entities:List[Entity],revision_nodes:List[RevisionNodes]...}
+    """
+
+    revised_entities, history = _modify_docproto(revision.entities)
+    return revised_entities, history
+
+
+def get_level(doc: DocumentWithRevisions):
+    try:
```

Comment:
Why does this need a try/except? could there be an if statement to check instead?

Reply:
If any error is raised for example if a node doesn't have a parent, if we used `if` it would raise an error and we want it to just return a `0`.

Reply:
If you are testing for whether the object has a `parent` attribute, you can call `hasattr(p, 'parent')`.  This could be just something like `while hasattr(p, 'parent'): level += 1; p = p.parent`